eras/mary/impl/src/Cardano/Ledger/Mary/Value.hs

```diff
@@ -286,64 +282,55 @@ roundupBytesToWords b = quot (b + wordLength - 1) wordLength
-        TypeUInt64 -> inject <$> decodeCoin "Coin in Value"
-        TypeListLen -> decodeValuePair decodeCoin
-        TypeListLen64 -> decodeValuePair decodeCoin
-        TypeListLenIndef -> decodeValuePair decodeCoin
-        _ -> fail $ "Value: expected array or int, got " ++ show tt
+decodeMaryValue = do
+  tt <- peekTokenType
+  case tt of
+    TypeUInt -> inject . Coin . toInteger <$> decodeWord64
+    TypeUInt64 -> inject . Coin . toInteger <$> decodeWord64
+    TypeListLen -> decodeValuePair (toInteger <$> decodeWord64)
+    TypeListLen64 -> decodeValuePair (toInteger <$> decodeWord64)
+    TypeListLenIndef -> decodeValuePair (toInteger <$> decodeWord64)
+    _ -> fail $ "MaryValue: expected array or int, got " ++ show tt
 
 decodeValuePair ::
   Crypto c =>
-  (forall t. String -> Decoder t Coin) ->
+  (forall t. Decoder t Integer) ->
   Decoder s (MaryValue c)
-decodeValuePair decodeAmount =
+decodeValuePair decodeMultiAssetAmount =
   decode $
     RecD MaryValue
-      <! D (decodeAmount "Coin in Value")
-      <! D (decodeMultiAssetMaps (unCoin <$> decodeAmount "MultiAsset in Value"))
-
-encodeMultiAssetMaps ::
-  Crypto c =>
-  MultiAsset c ->
-  Encoding
-encodeMultiAssetMaps (MultiAsset m) = encCBOR m
+      <! D (Coin . toInteger <$> decodeWord64)
+      <! D (decodeMultiAsset decodeMultiAssetAmount)
 
 -- | `MultiAsset` can be used in two different circumstances:
--- during minting and in `MaryValue` while sending.
--- In the latter case, MultiAsset amounts cannot be negative,
--- while in minting negative indicates burning, and should not be zero.
-decodeMultiAssetMaps :: Crypto c => (forall t. Decoder t Integer) -> Decoder s (MultiAsset c)
-decodeMultiAssetMaps decodeAmount = do
-  ma <- decodeMap decCBOR (decodeMap decCBOR decodeAmount)
-  if isMultiAssetSmallEnough (MultiAsset ma)
-    then
-      ifDecoderVersionAtLeast
+--
+-- 1. In `MaryValue` while sending, where amounts must be positive.
+-- 2. During minting, both negative and positive are allowed, but not zero.
+--
+-- In both cases MultiAsset cannot be too big for compact representation and it must not
+-- contain empty Maps.
+decodeMultiAsset :: Crypto c => (forall t. Decoder t Integer) -> Decoder s (MultiAsset c)
+decodeMultiAsset decodeAmount = do
+  ma <-
+    MultiAsset
+      <$> ifDecoderVersionAtLeast
         (natVersion @9)
-        (MultiAsset ma <$ forM_ ma (\m -> when (Map.null m) $ fail "Empty Assets are not allowed"))
-        (pure $ MultiAsset $ prune ma)
-    else fail "MultiAsset too big to compact"
-
-decodeMultiAssetMint :: Crypto c => Decoder s (MultiAsset c)
-decodeMultiAssetMint = decodeMultiAssetMaps decodeIntegerBounded64
+        decodeWithEnforcing
+        decodeWithPrunning
+  ma <$ unless (isMultiAssetSmallEnough ma) (fail "MultiAsset is too big to compact")
+  where
+    decodeWithEnforcing =
+      decodeMap decCBOR $ do
+        m <- decodeMap decCBOR $ do
+          amount <- decodeAmount
```

Comment:
Something I don't understand here: 
In the previous version, the value-Decoder we passed  to `decodeMap` (second argument) was `(decodeMap decCBOR decodeAmount)` , and after that we would do enforcing.  But in the new version, the decoder seems to be `decodeAmount` (plus the enforcing).  Are they equivalent ? (sorry if it's obvious, I'm probably just missing it)

Reply:
I am really sorry, I don't understand the question. It's a bit ambiguous. We are enforcing two things:
* Non-zero amounts
* Non-empty maps.

Which enforcement are you asking about?

Reply:
I'm sorry, I missed that there are two lines containing `decodeMap decCBOR $ do`, at 327 and 328, so I was under the impression that in the case of  `decodeWithEnforcing`, we are passing simply `decodeAmount` as a decoder for the value of the map, rather than `decodeMap decCBOR decodeAmount`.  Sorry!